src/Agent.Listener/CapabilitiesProvider.cs

```diff
@@ -0,0 +1,291 @@
+            new List<Capability>
+            {
+                new Capability( "ant" ),
+                new Capability( "bundler", "bundle" ),
+                new Capability( "clang" ),
+                new Capability( "cmake" ),
+                new Capability( "curl" ),
+                new Capability( "git" ),
+                new Capability( "gulp" ),
+                new Capability( "java" ),
+                new Capability( "JDK", "javac" ),
+                new Capability( "make" ),
+                new Capability( "maven", "mvn" ),
+                new Capability( "MSBuild", "xbuild" ),
+                new Capability( "node.js", "node" ),
+                new Capability( "node.js", "nodejs" ),
+                new Capability( "npm" ),
+                new Capability( "python" ),
+                new Capability( "python3" ),
+                new Capability( "sh" ),
+                new Capability( "subversion", "svn" ),
+                new Capability( "ruby" ),
+                new Capability( "rake" ),
+                new Capability( "Xamarin.iOS", "mdtool", new string[] { "/Applications/Xamarin Studio.app/Contents/MacOS/mdtool" } ),
+                new Capability( "Xamarin.Android", "mandroid", new string[] { "/Library/Frameworks/Xamarin.Android.framework/Commands/mandroid" } )
+            };
+
+        private static readonly List<ToolCapability> _toolCapabilities =
+            new List<ToolCapability>
+            {
+                new ToolCapability( "xcode", "xcode-select", "-p" ),
+            };
+
+        private static readonly string[] _ignoredEnvVariables = new string[] {
+            "TERM_PROGRAM",
+            "TERM",
+            "TERM_PROGRAM_VERSION",
+            "SHLVL",
+            "ls_colors",
+            "comp_wordbreaks"
+        };
+
+        // Ignore env vars specified in the 'VSO_AGENT_IGNORE' env var
+        private const string EnvIgnore = "VSO_AGENT_IGNORE";
+
+        private Dictionary<string, string> _capsCache;
+
+        public List<Capability> RegularCapabilities
+        {
+            get
+            {
+                return _regularCapabilities;
+            }
+        }
+
+        public List<ToolCapability> ToolCapabilities
+        {
+            get
+            {
+                return _toolCapabilities;
+            }
+        }
+
+        public async Task<Dictionary<string, string>> GetCapabilitiesAsync(string agentName, CancellationToken token)
+        {
+            if (_capsCache != null)
+            {
+                return new Dictionary<string, string>(_capsCache, StringComparer.OrdinalIgnoreCase);
+            }
+
+            var caps = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            GetFilteredEnvironmentVars(caps);
+
+            GetRegularCapabilities(caps, token);
+
+            await GetToolCapabilities(caps, token);
+
+            caps["Agent.Name"] = agentName ?? string.Empty;
+#if OS_LINUX
```

Comment:
Can we add the #if def to constants and make this just set Agent.OS via the const.  The const would be platform and would allow other code to switch on platform without doing if defs